detox/src/Detox.js

```diff
@@ -144,6 +120,70 @@ class Detox {
     await this._artifactsManager.onSuiteEnd(suite);
   }
 
+  async _doInit(userParams) {
+    const sessionConfig = await this._getSessionConfig();
+    const params = {
+      launchApp: true,
+      initGlobals: true,
+      ...userParams,
+    };
+
+    if (!this._userSession) {
+      this._server = new DetoxServer({
+        log: logger,
+        port: new URL(sessionConfig.server).port,
+      });
+    }
+
+    this._client = new Client(sessionConfig);
+    this._client.setNonresponsivenessListener(this._onNonresnponsivenessEvent.bind(this));
+    await this._client.connect();
+
+    const DeviceDriverClass = DEVICE_CLASSES[this._deviceConfig.type];
+    if (!DeviceDriverClass) {
+      throw new Error(`'${this._deviceConfig.type}' is not supported`);
+    }
+
+    const deviceDriver = new DeviceDriverClass({
+      client: this._client,
+    });
+
+    Object.assign(this, deviceDriver.matchers);
+
+    this.device = new Device({
```

Comment:
Assigning instances to properties as soon as they are constructed.
It makes it easier to work with artifacts in cleanup phase when init is not really done.